teuthology/scrape.py

```diff
@@ -0,0 +1,512 @@
+# Origin: https://github.com/jcsp/scrape/blob/master/scrape.py
+# Author: John Spray (github.com/jcsp)
+
+import difflib
+from errno import ENOENT
+from gzip import GzipFile
+import sys
+import os
+import yaml
+from collections import defaultdict
+import re
+import logging
+import subprocess
+
+log = logging.getLogger('scrape')
+log.addHandler(logging.StreamHandler())
+log.setLevel(logging.INFO)
+
+
+class Reason(object):
+    def get_description(self):
+        return self.description
+
+    def get_detail(self):
+        return None
+
+
+def grep(path, expr):
+    """
+    Call out to native grep rather than feeding massive log files through python line by line
+    """
+    p = subprocess.Popen(["grep", expr, path], stdout=subprocess.PIPE)
+    out, err = p.communicate()
+    if p.returncode == 0:
+        return out.split("\n")
```

Comment:
Not sure this code is py3 compatible.

Reply:
Adding a `p.wait()` sets `p.returncode` to 0 (if grep process was successful). Tested this in both py2 and py3. Otherwise line 34 evaluates to `True` and `False` in py2 and py3 respectively.

Reply:
@grapheo12 the question is about `out.split("\n")`, since Popen will return Bytes in out but in python3 str should be prepended with `b` to compare with bytes.

Reply:
In that case, let me introduce a check:

```python
if type(out) != str:
    out = out.decode()
```

This should take care of the bytes vs str problem.

Reply:
We have a uniform practice for this kind of check to use `six.ensure_str`.